backends/ebpf/psa/externs/ebpfPsaDigest.cpp

```diff
@@ -0,0 +1,177 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ebpfPsaDigest.h"
+
+#include "backends/ebpf/ebpfType.h"
+#include "backends/ebpf/psa/ebpfPsaDeparser.h"
+
+namespace EBPF {
+
+class EBPFDigestPSAValueVisitor : public CodeGenInspector {
+ protected:
+    cstring digestMapInstance;
+    CodeGenInspector *codegen;
+    EBPFType *valueType;
+
+ public:
+    EBPFDigestPSAValueVisitor(P4::ReferenceMap *refMap, P4::TypeMap *typeMap,
+                              cstring digestMapInstance, CodeGenInspector *codegen,
+                              EBPFType *valueType)
+        : CodeGenInspector(refMap, typeMap),
+          digestMapInstance(digestMapInstance),
+          codegen(codegen),
+          valueType(valueType) {}
+
+    void pushExistingElem(const IR::Expression *expr) {
+        builder->emitIndent();
+        builder->appendFormat("bpf_map_push_elem(&%s, &", digestMapInstance);
+        codegen->visit(expr);
+        builder->append(", BPF_EXIST)");
+        builder->endOfStatement(true);
+    }
+
+    // handle expression like: "digest.pack(msg)"
+    bool preorder(const IR::PathExpression *pe) override {
+        pushExistingElem(pe);
+        return false;
+    }
+
+    // handle expression like: "digest.pack(metadata.msg)"
+    bool preorder(const IR::Member *member) override {
+        pushExistingElem(member);
+        return false;
+    }
+
+    // handle expression like: "digest.pack(value)"
+    bool preorder(const IR::Constant *c) override {
+        cstring tmpVar = refMap->newName("digest_entry");
+
+        builder->emitIndent();
+        builder->blockStart();
+
+        builder->emitIndent();
+        valueType->declare(builder, tmpVar, false);
+        builder->append(" = ");
+        codegen->visit(c);
+        builder->endOfStatement(true);
+
+        builder->emitIndent();
+        builder->appendFormat("bpf_map_push_elem(&%s, &%s, BPF_EXIST)", digestMapInstance, tmpVar);
+        builder->endOfStatement(true);
+
+        builder->blockEnd(true);
+        return false;
+    }
+
+    // handle expression like: "digest.pack({ value1, value2 })"
+    bool preorder(const IR::StructExpression *se) override {
+        cstring tmpVar = refMap->newName("digest_entry");
+
+        builder->emitIndent();
+        builder->blockStart();
+
+        builder->emitIndent();
+        valueType->declare(builder, tmpVar, false);
+        builder->endOfStatement(true);
+
+        for (const auto *f : se->components) {
+            unsigned width = EBPFInitializerUtils::ebpfTypeWidth(typeMap, f->expression);
+            builder->emitIndent();
+
+            if (EBPFScalarType::generatesScalar(width)) {
```

Comment:
I wonder whether this code pattern shouldn't be some utility function.

Reply:
I've added such function in `CodeGenInspector` which is used also in `preorder(const IR::AssignmentStatement)`